### ISSUE TITLE
Losetup device issues in post/pre processing.

### ISSUE THREAD

**Issue by wajihyassine** [google/turbinia#1122]
As I was running some load tests, I discovered an issue with the Losetup devices not being deleted properly and may be either an issue with how partitions are being mounted or removed, either due to the PartitionEnumeration code or due to an unknown error. What makes it even more difficult is that so far, I've only been able to reproduce the error when the whole infrastructure is deployed in GKE, with running Worker pods that get scaled and descaled and unable to put a debugger on where the error is happening. 

Sample of errors observed
```
"Evidence post-processing for test-disk-25gb-1:/p15 failed: Could not delete losetup device /dev/loop1"
"Evidence post-processing for test-disk-25gb-4:/p1 failed: Could not delete losetup device /dev/loop1"
"Evidence post-processing for test-disk-25gb-3:/p15 failed: Could not delete losetup device /dev/loop1"
"Evidence post-processing for test-disk-25gb-1:/p1 failed: Could not delete losetup device /dev/loop1"
```

**Comment by wajihyassine** [comment]
Getting some help from @aarontp, recommended that I try adding some Retry logic when the block device is being mounted to confirm that it is a block device. 

Added `IsBlockDevice` in `processors.mount_local.py`

```
def IsBlockDevice(path):
  """Checks path to determine whether it is a block device.

  Args:
      path: String of path to check.

  Returns:
      Bool indicating success.
  """
  if not os.path.exists(path):
    return False
  mode = os.stat(path).st_mode
  return stat.S_ISBLK(mode)
``` 

and retry logic in `processors.mount_local.PreprocessLosetup`
```
    # Make sure we have a proper block device
    for _ in range(RETRY_MAX):
      if IsBlockDevice(losetup_device):
        log.info(
            'Block device {0:s} successfully attached'.format(losetup_device))
        break
      if os.path.exists(losetup_device):
        log.info(
            'Block device {0:s} mode is {1}'.format(
                losetup_device,
                os.stat(losetup_device).st_mode))
      time.sleep(1)
    message = None
    if not os.path.exists(losetup_device):
      message = 'Device path {0:s} does not exist'.format(losetup_device)
    elif not IsBlockDevice(losetup_device):
      message = 'Device path {0:s} is not a block device'.format(losetup_device)
      if message:
        log.error(message)
        raise TurbiniaException(message)
```
 After running some tests, the above code doesn't seem to properly check if the device block is actually attached (due to seeing some errors trying to access the block devices manually)

**Comment by wajihyassine** [comment]
@aarontp recommended to try calling `losetup -a` instead for checking if the Device has been properly attached.

In `processors.mount_local.py` replaced the `IsBlockDevice` function with `IsLosetup`
```
def IsLosetup(path):
  # Check that the device was attached
  losetup_cmd = ['sudo', 'losetup', '-a']
  log.info('Running: {0:s}'.format(' '.join(losetup_cmd)))
  try:
    output = subprocess.check_output(losetup_cmd)
  except subprocess.CalledProcessError as e:
    raise TurbiniaException(
        'Could not check losetup device status {0!s}'.format(e))
  if output.find(path.encode('utf-8')) != -1:
    return False
  log.info('losetup device [{0!s}] deleted.'.format(path))
  return True
```

Updated `processors.mount_local.PreprocessLosetup` to call `IsLosetup` instead
```
    # Make sure we have a proper block device
    for _ in range(RETRY_MAX):
      if IsLosetup(losetup_device):
        log.info(
            'Block device {0:s} successfully attached'.format(losetup_device))
        break
      if os.path.exists(losetup_device):
        log.info(
            'Block device {0:s} mode is {1}'.format(
                losetup_device,
                os.stat(losetup_device).st_mode))
      time.sleep(1)
    message = None
    if not os.path.exists(losetup_device):
      message = 'Device path {0:s} does not exist'.format(losetup_device)
    elif not IsLosetup(losetup_device):
      message = 'Device path {0:s} is not a block device'.format(losetup_device)
      if message:
        log.error(message)
        raise TurbiniaException(message)
``` 

After running some tests, there still seems to be an issue with detaching some of the loop devices by partition.

**Comment by wajihyassine** [comment]
@aarontp next suggestion to also try adding the retry logic for when the loopback device is being detached to ensure that there is enough time for it to detach.

```
    losetup_cmd = ['sudo', 'losetup', '-d', device_path]
    for _ in range(RETRY_MAX):
      log.info('Running: {0:s}'.format(' '.join(losetup_cmd)))
      try:
        subprocess.check_call(losetup_cmd)
      except subprocess.CalledProcessError as e:
        turbinia_failed_loop_device_detach.inc()
        raise TurbiniaException(
            'Could not delete losetup device {0!s}'.format(e))
      # Check that the device was actually removed
      losetup_cmd = ['sudo', 'losetup', '-a']
      log.info('Running: {0:s}'.format(' '.join(losetup_cmd)))
      try:
        output = subprocess.check_output(losetup_cmd)
      except subprocess.CalledProcessError as e:
        raise TurbiniaException(
            'Could not check losetup device status {0!s}'.format(e))
      if output.find(device_path.encode('utf-8')) != -1:
        log.info(
            'Block device {0:s} mode is {1}'.format(
                device_path,
                os.stat(device_path).st_mode))
        else:
           break
      time.sleep(1)
    # Check if device failed.
    if output.find(device_path.encode('utf-8')) != -1:
      turbinia_failed_loop_device_detach.inc()
      raise TurbiniaException(
          'Could not delete losetup device {0!s}'.format(device_path))
    log.info('losetup device [{0!s}] deleted.'.format(device_path))
```

I ran some tests to see if the preprocess logic properly detected that the loop device is not a block device. 
```
Error running preprocessor for test-disk-25gb-15:/p1: Device path /dev/loop101 is not a block device
Device path /dev/loop101 is not a block device
Error running preprocessor for test-disk-25gb-15:/p1: Device path /dev/loop100 is not a block device
Device path /dev/loop100 is not a block device
```

I no longer see the `Could not delete losetup device` error message indicating that the preprocess is likely failing before the Task runs (given not equivalent device). However now seeing more failed Tasks as a result.
```
* YaraAnalysisTask (test-disk-25gb-6:/p1): Task Result was auto-closed from task executor on turbinia-worker-5899cf6d9c-48wkm likely due to previous failures.  Previous status: [YaraAnalysisTask Task failed with exception: [Evidence DiskPartition:test-disk-25gb-6:/p1:None being processed by Task YaraAnalysisTask requires Evidence to be in state ATTACHED, but earlier pre-processors may have failed.  Current state is [MOUNTED: False, ATTACHED: False, DECOMPRESSED: False, CONTAINER_MOUNTED: False]. See previous logs for more information.]]
--
* DockerContainersEnumerationTask (test-disk-25gb-7:/p1): Task Result was auto-closed from task executor on turbinia-worker-5899cf6d9c-gtwjd likely due to previous failures.  Previous status: [DockerContainersEnumerationTask Task failed with exception: [Evidence DiskPartition:test-disk-25gb-7:/p1:None being processed by Task DockerContainersEnumerationTask requires Evidence to be in state ATTACHED, but earlier pre-processors may have failed.  Current state is [MOUNTED: False, ATTACHED: False, DECOMPRESSED: False, CONTAINER_MOUNTED: False]. See previous logs for more information.]]
--
* StringsAsciiTask (test-disk-25gb-8:/p15): Task Result was auto-closed from task executor on turbinia-worker-5899cf6d9c-mcp5j likely due to previous failures.  Previous status: [StringsAsciiTask Task failed with exception: [Evidence DiskPartition:test-disk-25gb-8:/p15:None being processed by Task StringsAsciiTask requires Evidence to be in state ATTACHED, but earlier pre-processors may have failed.  Current state is [MOUNTED: False, ATTACHED: False, DECOMPRESSED: False, CONTAINER_MOUNTED: False]. See previous logs for more information.]]
--
FsstatTask (test-disk-25gb-9:/p1): Task Result was auto-closed from task executor on turbinia-worker-5899cf6d9c-tzflh likely due to previous failures.  Previous status: [FsstatTask Task failed with exception: [Evidence DiskPartition:test-disk-25gb-9:/p1:None being processed by Task FsstatTask requires Evidence to be in state ATTACHED, but earlier pre-processors may have failed.  Current state is [MOUNTED: False, ATTACHED: False, DECOMPRESSED: False, CONTAINER_MOUNTED: False]. See previous logs for more information.]]
```

**Comment by wajihyassine** [comment]
@aarontp this is where I got stuck, if you have time and can PTAL per our chats around this.

**Comment by wajihyassine** [comment]
Part 1. temp fix in #1144 and part 2. fix related to race condition in #1145
